libraries/lib-transactions/TransactionScope.cpp

```diff
@@ -0,0 +1,101 @@
+/*!********************************************************************
+
+  Audacity: A Digital Audio Editor
+
+  @file TransactionScope.cpp
+
+  Paul Licameli split from DBConnection.cpp
+
+**********************************************************************/
+
+#include "TransactionScope.h"
+#include "InconsistencyException.h"
+#include <wx/log.h>
+
+namespace {
+TransactionScope::TransactionScopeImplFactory &GetFactory()
+{
+   static TransactionScope::TransactionScopeImplFactory theFactory;
+   return theFactory;
+}
+
+TransactionScope::AutoSaveFunction &GetAutoSave()
+{
+   static TransactionScope::AutoSaveFunction theFunction;
+   return theFunction;
+}
+}
+
+auto TransactionScope::InstallImplementation(
+   TransactionScopeImplFactory factory) -> TransactionScopeImplFactory
+{
+   auto &theFactory = GetFactory();
+   auto result = theFactory;
+   theFactory = move(factory);
+   return result;
+}
+
+auto TransactionScope::InstallAutoSave(
+   AutoSaveFunction function) -> AutoSaveFunction
+{
+   auto &theFunction = GetAutoSave();
+   auto result = theFunction;
+   theFunction = move(function);
+   return theFunction;
+}
+
+void TransactionScope::AutoSave(AudacityProject &project)
+{
+   auto &function = GetAutoSave();
+   if (function)
+      function(project);
+}
+
+TransactionScopeImpl::~TransactionScopeImpl() = default;
+
+TransactionScope::TransactionScope(
+   AudacityProject &project, const char *name)
+:  mName(name)
+{
+   auto &factory = GetFactory();
+   if (!factory)
+      return;
+   mpImpl = factory(project);
+   if (!mpImpl)
+      return;
+
+   mInTrans = mpImpl->TransactionStart(mName);
+   if ( !mInTrans )
+      // To do, improve the message
+      throw SimpleMessageBoxException( ExceptionType::Internal,
+         XO("Database error.  Sorry, but we don't have more details."), 
```

Comment:
I have a strong opinion, that implementation should throw or otherwise provide the user with details on what to do, rather than a generic message that something has failed. The user side of error handling is already way too confusing and unhelpful for the user

Reply:
A fair point but a separate question from the proposal in this PR.  This much of exsitent behavior is preserved without change, though moved into another file.

Reply:
I understand, that this somewhat maintains the current behavior. My point is - such separation will make it hard to fix it in the future.